File: packages/arabic_bidi/dart/lib/src/logger.dart
library;

import 'terminal.dart';

/// Arabic-aware terminal logger.
///
/// Applies [prepareForTerminal] to every message before printing.
/// Drop-in replacement for bare [print] calls in CLI and script environments.
///
/// Output format: `[timestamp?][prefix?][LEVEL] message`
class ArabicLogger {
  /// Optional label prepended to every line as `[prefix]`.
  final String? prefix;

  /// When true, prepends an ISO-8601 timestamp `[yyyy-MM-dd HH:mm:ss]`.
  final bool useTimestamp;

  /// Terminal preparation options applied to every message.
  final ArabicTerminalOptions terminalOptions;

  const ArabicLogger({
    this.prefix,
    this.useTimestamp = false,
    this.terminalOptions = const ArabicTerminalOptions(),
  });

  void info(String message) => _log('INFO', message);
  void error(String message) => _log('ERROR', message);
  void warn(String message) => _log('WARN', message);
  void debug(String message) => _log('DEBUG', message);

  void _log(String level, String message) {
    final prepared = prepareForTerminal(message, options: terminalOptions);
    final sb = StringBuffer();
    if (useTimestamp) {
      final now = DateTime.now();
      final y = now.year.toString().padLeft(4, '0');
      final mo = now.month.toString().padLeft(2, '0');
      final d = now.day.toString().padLeft(2, '0');
      final h = now.hour.toString().padLeft(2, '0');
      final mi = now.minute.toString().padLeft(2, '0');
      final s = now.second.toString().padLeft(2, '0');
      sb.write('[$y-$mo-$d $h:$mi:$s]');
    }
    if (prefix != null) sb.write('[$prefix]');
    sb.write('[$level] $prepared');
    print(sb.toString());
  }
}

/// Ready-to-use default logger instance.
const arabicLogger = ArabicLogger();

File: packages/arabic_bidi/js/src/logger.ts
// Arabic-aware terminal logger — devsamhan-arabic spec v1.0.0
//
// Output format: [timestamp?][prefix?][LEVEL] message

import { prepareForTerminal, TerminalOptions } from './terminal';

export interface ArabicLoggerOptions {
  prefix?: string;
  useTimestamp?: boolean;
  terminalOptions?: TerminalOptions;
}

export class ArabicLogger {
  private readonly prefix: string | undefined;
  private readonly useTimestamp: boolean;
  private readonly terminalOptions: TerminalOptions;

  constructor(options?: ArabicLoggerOptions) {
    this.prefix = options?.prefix;
    this.useTimestamp = options?.useTimestamp ?? false;
    this.terminalOptions = options?.terminalOptions ?? {};
  }

  info(message: string): void  { this._log('INFO',  message); }
  error(message: string): void { this._log('ERROR', message); }
  warn(message: string): void  { this._log('WARN',  message); }
  debug(message: string): void { this._log('DEBUG', message); }

  private _log(level: string, message: string): void {
    const prepared = prepareForTerminal(message, this.terminalOptions);
    let line = '';
    if (this.useTimestamp) {
      const now = new Date();
      const pad = (n: number) => String(n).padStart(2, '0');
      const ts = `${now.getFullYear()}-${pad(now.getMonth() + 1)}-${pad(now.getDate())} ${pad(now.getHours())}:${pad(now.getMinutes())}:${pad(now.getSeconds())}`;
      line += `[${ts}]`;
    }
    if (this.prefix !== undefined) line += `[${this.prefix}]`;
    line += `[${level}] ${prepared}`;
    console.log(line);
  }
}

export const arabicLogger = new ArabicLogger();
